habitat_code/nobodyliveshere-kmeans.py

```diff
@@ -0,0 +1,804 @@
+if 'processed_data' not in granule_gdf.columns:
+    granule_gdf['processed_data'] = pd.Series(dtype=object)
+
+# Function to process a single band
+@cached(lambda granule_id, band_key, band_url, mask_bits, scale_factor, mask, watershed: f"{granule_id}-{band_key}")
+def process_band(granule_iid, band_key, band_path, mask_bits, scale_factor, mask,  watershed):
+    # Load the band data
+    band_da = rxr.open_rasterio(band_path) * scale_factor
+
+    # Ensure the watershed is in the same CRS as the band data
+    watershed_reprojected = watershed.to_crs(band_da.rio.crs)
+
+    # Clip and reproject the band data to the watershed
+    band_clipped = band_da.rio.clip(watershed_reprojected.geometry, watershed_reprojected.crs)
+
+    # Load, clip, and reproject the Fmask data to the watershed
+    fmask_da = rxr.open_rasterio(fmask_path)
+    fmask_clipped = fmask_da.rio.clip(watershed_reprojected.geometry, watershed_reprojected.crs)
+    print(fmask_clipped.shape)
+
+    # Ensure the mask is aligned with the band data
+    mask = np.unpackbits(fmask_clipped.astype(np.uint8)).reshape(fmask_clipped.shape + (-1,))
+    mask_aligned = np.prod(mask[..., mask_bits] == 0, axis=-1)
+    if mask_aligned.shape != band_clipped.shape:
+        raise ValueError("The mask and band data are not aligned. Check the spatial extents and resolutions.")
+
+    # Apply the mask to the clipped and aligned band data
+    masked_band_da = band_clipped.where(mask_aligned, other=np.nan)
+
+    return masked_band_da
+
+
+
+
+# Initialize a counter
+counter = 0
+# Set the number of rows to process (change this number as needed)
+num_rows_to_process = len(granule_gdf)
+
+# Iterate over each group and process the bands
+for granule_id, group in granule_gdf.groupby('granule_id'):
+    print(f"Processing granule: {granule_id}")
+    fmask_path = group.iloc[0]['fmask_url']
+    fmask_da = rxr.open_rasterio(fmask_path)
+    bits = np.unpackbits(fmask_da.astype(np.uint8)).reshape(fmask_da.shape + (-1,))
+    mask = np.prod(bits[..., mask_bits] == 0, axis=-1)
+
+    for _, row in group.iterrows():
+        if counter >= num_rows_to_process:
+            break  # Stop processing if the counter reaches the limit
+
+        band_number = row['band_number']
+        band_url = row['band_url']
+        if isinstance(band_number, float) and not np.isnan(band_number):
+            band_number = int(band_number)
+        if isinstance(band_number, int) and band_url.endswith('.tif') and 'B' in band_url:
+            band_key = 'B' + str(band_number)
+            processed_data = process_band(granule_id, band_key, band_url, mask_bits, scale_factor, mask, watershed)
+            granule_gdf.at[_, 'processed_data'] = processed_data
+            print(f"Processed data for {granule_id} - {band_key}: {granule_gdf.at[_, 'processed_data']}")
+
+        counter += 1  # Increment the counter after processing each row
+
+        if counter >= num_rows_to_process:

```

Comment:
Interesting way of doing this, cool!

Reply:
I had set it up that way to test out different numbers of granules because of the amount of trouble I had processing the individual ones. That way I wouldn't have to create another object or break  a for loop.